File: app/src/main/java/com/shicheeng/copymanga/data/MangaReadJson.java
package com.shicheeng.copymanga.data;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.shicheeng.copymanga.KeyWordSwap;
import com.shicheeng.copymanga.apiName;

import java.io.IOException;
import java.util.HashMap;
import java.util.Objects;
import java.util.concurrent.ExecutionException;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

public class MangaReadJson {

    public static HashMap<Integer, String> getPicBitmap(String pathWord, String uuid)
            throws IOException, ExecutionException, InterruptedException {
        String url = apiName.mangaPhotoApi(pathWord, uuid);

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder().removeHeader(KeyWordSwap.USER_AGENT_WORD)
                .addHeader(KeyWordSwap.USER_AGENT_WORD, KeyWordSwap.FAKE_USER_AGENT)
                .url(url)
                .build();

        try (Response response = client.newCall(request).execute()) {
            HashMap<Integer, String> hashMap = new HashMap<>();
            String json = Objects.requireNonNull(response.body()).string();
            JsonObject object = JsonParser.parseString(json).getAsJsonObject().getAsJsonObject("results");
            JsonArray array = object.get("chapter").getAsJsonObject().getAsJsonArray("contents");
            JsonArray array2 = object.get("chapter").getAsJsonObject().getAsJsonArray("words");
            for (int i = 0; i < array.size(); i++) {
                String url2 = array.get(i).getAsJsonObject().get("url").getAsString();
                int num = array2.get(i).getAsInt();
                hashMap.put(num, url2);
            }
            return hashMap;
        }

    }



    public static String getNextChapterName(String path_word, String nextUuid) throws IOException {
        String url = apiName.mangaPhotoApi(path_word, nextUuid);

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder().removeHeader(KeyWordSwap.USER_AGENT_WORD)
                .addHeader(KeyWordSwap.USER_AGENT_WORD, KeyWordSwap.FAKE_USER_AGENT)
                .url(url)
                .build();

        try (Response response = client.newCall(request).execute()) {
            String json = Objects.requireNonNull(response.body()).string();
            JsonObject object = JsonParser.parseString(json).getAsJsonObject().getAsJsonObject("results");

            return object.get("chapter").getAsJsonObject().get("name").getAsString();
        }
    }

    public static String getNextUUID(String pathWord, String uuid) throws IOException {
        String url = apiName.mangaPhotoApi(pathWord, uuid);

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder().removeHeader(KeyWordSwap.USER_AGENT_WORD)
                .addHeader(KeyWordSwap.USER_AGENT_WORD, KeyWordSwap.FAKE_USER_AGENT)
                .url(url)
                .build();

        try (Response response = client.newCall(request).execute()) {
            String json = Objects.requireNonNull(response.body()).string();
            JsonObject object = JsonParser.parseString(json).getAsJsonObject().getAsJsonObject("results");
            String uuid2;
            if (!object.get("chapter").getAsJsonObject().get("next").isJsonNull()) {
                uuid2 = object.get("chapter").getAsJsonObject().get("next").getAsString();
            } else {
                uuid2 = KeyWordSwap.NON_JSON;
            }

            return uuid2;
        }
    }

}

File: app/src/main/java/com/shicheeng/copymanga/json/MangaReadJson.kt
package com.shicheeng.copymanga.json

import com.google.gson.JsonParser
import com.shicheeng.copymanga.data.MangaReadInformation
import com.shicheeng.copymanga.data.MangaReaderPage
import com.shicheeng.copymanga.util.ApiName
import com.shicheeng.copymanga.util.FileUtil
import com.shicheeng.copymanga.util.KeyWordSwap
import com.shicheeng.copymanga.util.OkhttpHelper
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Request

object MangaReadJson {


    suspend fun getMangaInfo(
        pathWord: String?,
        uuid: String?,
        fileUtil: FileUtil,
    ): MangaReadInformation = withContext(Dispatchers.Default) {
        if (fileUtil.isChapterDownloadedWithStringList(pathWord, uuid)) {
            val json = fileUtil.findDownloadChapterInfo(pathWord, uuid)
            val time = json?.get("chapter_name")?.asString
            val listSize = fileUtil.ifChapterDownloaded(pathWord, uuid).size
            MangaReadInformation(null, time, listSize)
        } else {
            val url = ApiName.mangaPhotoApi(pathWord, uuid)
            val client = OkhttpHelper.getInstance()
            val request: Request = Request.Builder().removeHeader(KeyWordSwap.USER_AGENT_WORD)
                .addHeader(KeyWordSwap.USER_AGENT_WORD, KeyWordSwap.FAKE_USER_AGENT)
                .headers(MangaInfoJson.headers)
                .url(url)
                .build()
            client.newCall(request).execute().use { response ->
                val json = response.body?.string()
                val jsonObject =
                    JsonParser.parseString(json).asJsonObject.getAsJsonObject("results")
                val jsonChapter = jsonObject["chapter"].asJsonObject
                val time = jsonChapter["name"].asString //This is manga chapter name. I was wrong.
                val subtitle = jsonChapter["datetime_created"].asString
                val size = jsonChapter["size"].asInt
                MangaReadInformation(subtitle, time, size)
            }

        }
    }

    suspend fun getMangaPic(
        pathWord: String?,
        uuid: String?,
        fileUtil: FileUtil,
    ): List<MangaReaderPage> =
        withContext(Dispatchers.Default) {
            if (fileUtil.isChapterDownloadedWithStringList(pathWord, uuid)) {
                val list = fileUtil.ifChapterDownloaded(pathWord, uuid)
                val sortedList = list.sortedWith { text1, text2 ->
                    text1.url.split("/").last().split(".").first().toInt()
                        .compareTo(text2.url.split("/").last().split(".").first().toInt())
                }
                val newList = buildList {
                    for (i in sortedList.indices) {
                        add(sortedList[i].copy(index = i))
                    }
                }
                newList
            } else {
                val url = ApiName.mangaPhotoApi(pathWord, uuid)
                val client = OkhttpHelper.getInstance()
                val request: Request = Request.Builder().removeHeader(KeyWordSwap.USER_AGENT_WORD)
                    .addHeader(KeyWordSwap.USER_AGENT_WORD, KeyWordSwap.FAKE_USER_AGENT)
                    .headers(MangaInfoJson.headers)
                    .url(url)
                    .build()
                client.newCall(request).execute().use { response ->
                    val hashMap = HashMap<Int, String>()
                    val json = response.body?.string()
                    val jsonObject =
                        JsonParser.parseString(json).asJsonObject.getAsJsonObject("results")
                    val jsonChapter = jsonObject["chapter"].asJsonObject
                    val array = jsonChapter.getAsJsonArray("contents")
                    val array2 = jsonChapter.getAsJsonArray("words")
                    for (i in 0 until array.size()) {
                        val url2 = array[i].asJsonObject["url"].asString
                        val num = array2[i].asInt
                        hashMap[num] = url2
                    }
                    val listUrl = buildList {
                        for (i in 0 until hashMap.size) {
                            val urlImage = hashMap[i]
                            urlImage?.let {
                                add(MangaReaderPage(it, uuid, i))
                            }
                        }
                    }
                    listUrl
                }
            }

        }
}